tough/src/http.rs

```diff
@@ -0,0 +1,256 @@
+use crate::error::{self, Error, Result};
+use crate::Transport;
+use reqwest::blocking::{Client, ClientBuilder, Request, Response};
+use reqwest::header::{self, HeaderValue, ACCEPT_RANGES};
+use reqwest::Method;
+use snafu::ResultExt;
+use std::cmp::Ordering;
+use std::io::Read;
+use std::time::Duration;
+use url::Url;
+
+/// Settings for the HTTP client including retry strategy and timeouts.
+#[derive(Clone, Copy, Debug)]
+pub struct ClientSettings {
+    /// Set a timeout for connect, read and write operations.
+    pub timeout: Duration,
+    /// Set a timeout for only the connect phase.
+    pub connect_timeout: Duration,
+    /// The total number of times we will try to get the response.
+    pub tries: u32,
+    /// The pause between the first and second try.
+    pub initial_backoff: Duration,
+    /// The maximum length of a pause between retries.
+    pub max_backoff: Duration,
+    /// The exponential backoff factor, the factor by which the pause time will increase after each
+    /// try until reaching `max_backoff`.
+    pub backoff_factor: f32,
+}
+
+impl Default for ClientSettings {
+    fn default() -> Self {
+        Self {
+            timeout: std::time::Duration::from_secs(30),
+            connect_timeout: std::time::Duration::from_secs(10),
+            /// try / 100ms / try / 150ms / try / 220ms / try
+            tries: 4,
+            initial_backoff: std::time::Duration::from_millis(100),
+            max_backoff: std::time::Duration::from_secs(1),
+            backoff_factor: 1.5,
+        }
+    }
+}
+
+/// An HTTP `Transport` with retry logic.
+#[derive(Clone, Copy, Debug)]
+pub struct HttpTransport {
+    settings: ClientSettings,
+}
+
+impl Default for HttpTransport {
+    fn default() -> Self {
+        HttpTransport::from_settings(ClientSettings::default())
+    }
+}
```

Comment:
This can be `#[derive]`d.

Reply:
[2a161fa](https://github.com/awslabs/tough/compare/7a2368bcfd892bbc54f3e89a9767dcf15aa2c9ce..2a161faf7544128fe1ca1ea4a6b9339d41e8259c)